pygmt/helpers/utils.py

```diff
@@ -58,26 +63,43 @@ def data_kind(data, x=None, y=None, z=None, required_z=False):
     >>> data_kind(data=np.arange(10).reshape((5, 2)), x=None, y=None)
     'matrix'
     >>> data_kind(data="my-data-file.txt", x=None, y=None)
-    'file'
+    'file_or_arg'
     >>> data_kind(data=pathlib.Path("my-data-file.txt"), x=None, y=None)
-    'file'
+    'file_or_arg'
+    >>> data_kind(data=None, x=None, y=None, optional_data=True)
+    'file_or_arg'
+    >>> data_kind(data=2.0, x=None, y=None, optional_data=True)
+    'file_or_arg'
+    >>> data_kind(data=True, x=None, y=None, optional_data=True)
+    'file_or_arg'
     >>> data_kind(data=xr.DataArray(np.random.rand(4, 3)))
     'grid'
     """
-    if data is None and x is None and y is None:
-        raise GMTInvalidInput("No input data provided.")
-    if data is not None and (x is not None or y is not None or z is not None):
-        raise GMTInvalidInput("Too much data. Use either data or x and y.")
-    if data is None and (x is None or y is None):
+    # pylint: disable=too-many-branches
+    # validate the combinations of data/x/y/z
```

Comment:
> but I also find that codes in data_kind is not that easy to understand. So better ideas about data_kind are welcomed.

How about we split the function into two? A `_validate_data_input` for the first half that checks that there's no missing or invalid data, and then the second half can stay in `data_kind` and that checks the instance type of the dataset.

Reply:
Sounds a good idea. Will do later.